### PR TITLE
Add normative DID URL Dereferencing section.

### PR DESCRIPTION
**Pull request by msporny**
This adds the normative text related to DID URL Dereferencing.

This is a WIP, do not merge.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/264.html" title="Last updated on May 14, 2020, 2:41 PM UTC (a99179c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/264/5c839b1...a99179c.html" title="Last updated on May 14, 2020, 2:41 PM UTC (a99179c)">Diff</a>

### CODE REVIEW

**Comment by philarcher** [comment]
Same comment as for PR 263. This:

"The general process of DID URL dereferencing is described in § 10.2 DID URL Dereferencing . The details of implementing a DID URL dereferencer are described in the DID Resolution specification [DID-RESOLUTION]."

is out of order IMO. From a normative POV, the DID-RESOLUTION doc does not exist. I know it does and I dearly wish it were in scope of the WG - that would make life a lot easier, but it isn't. So we can't treat it as if it is.

Again, write the normative sections without any ref to DID-RESOLUTION, then create one simple, short informative section in the doc, somewhere, pointing to it.

**Comment by jandrieu** [comment]
IMO, dereferencing is out of scope. Resolving a DID to a DID Document is one thing--that's the whole point of the architecture: separating the identifier from associated cryptographic material. 

However, once the spec moves into how we use that DID Document to dereference a specific resource OTHER than the DID Document, it extends beyond the minimum necessary for decentralized identifiers, with privacy implications that I believe are currently underexamined or dismissed to date.

**Comment by iherman** [comment]
> Same comment as for PR 263. This:
>
> "The general process of DID URL dereferencing is described in § 10.2 DID URL Dereferencing . The details of implementing a DID URL dereferencer are described in the DID Resolution specification [DID-RESOLUTION]."

@philarcher where did you find this quote in this PR? The "style" of this PR text is something like this quote:

>  The DID resolution function resolves a DID into a DID document by using the "Read" operation of the applicable DID method. (See § 9.2.2 Read/Verify .) The details of how this process is accomplished is outside the scope of this specification, but all implementations MUST implement a function in the form:
>
> resolve ( did, input-metadata ) -> ( did-document, did-document-metadata, did-resolution-metadata ) 

Which sounds perfectly fine for me in separating what is normative and what is not.

**Comment by iherman** [comment]
> However, once the spec moves into how we use that DID Document to dereference a specific resource OTHER than the DID Document, it extends beyond the minimum necessary for decentralized identifiers, with privacy implications that I believe are currently underexamined or dismissed to date.

The only logical conclusion from this is that the spec MUST NOT include the DID URL as a notion. This is a consistent view but, at this moment, we do have this notion. If we have, then we are supposed to define what it is...

**Comment by peacekeeper** [comment]
@jandrieu are you also objecting to having DID URLs in the spec? Or are you fine with having DID URLs, but you just object to talking about dereferencing them?

**Comment by jandrieu** [comment]
@peacekeeper That's a great question.

My preference would be a much stripped down spec, with a MVP that cleanly and clearly supports decentralized proof of control of identifiers. The mapping of those identifiers to resources and verification methods is, IMO, better handled at a different level, mainly because of concerns over privacy implications of current approaches. 

That said, I expect that some way to go from a DID to a particular resource (treating a DID as a URL) is going to be important to too many participants and we should figure that out. I just don't think we've figured it out yet. In particular, I believe the fragility caused by merging parameters to be so confusing to most developers that many who experiment with DIDs will find they don't work very well--in the same sense that XHTML didn't work very well because it was overfragile.

So, clarifying that resolution means returning a DID Document, to me, is absolutely in scope.

What is done with that DID Document to get to a non-DID Document resource is a bigger problem that we have not yet figured out and many of the proposed solutions give me considerable pause because of layer violations, such as multiple service endpoints and embedding resources in the DID Document.

IMO, we don't need these features to ship a viable solution for DIDs that enable cryptographic proof of control without relying on a trusted third party.

So, IMO, resolution is in scope for DID-Core, but I'm struggling to see how dereferencing is--AND I see more problems than consensus in the myriad of options currently under discussion.

**Comment by peacekeeper** [comment]
@jandrieu to follow up on this thread, sorry it took so long to respond

> That said, I expect that some way to go from a DID to a particular resource (treating a DID as a URL) is going to be important to too many participants and we should figure that out. I just don't think we've figured it out yet.

I completely agree that we haven't figured out all the details of what a DID URL can potentially dereference to, but I also think that there is overwhelming support for having DID URLs per se (at a minimum, super-simple DID URLs such as `did:ex:123#key-1`). Therefore, I think it is only logical that we define the dereferencing contract in addition to the resolution contract, would you agree?

However, having said that..

> What is done with that DID Document to get to a non-DID Document resource is a bigger problem

I agree with your concerns that we haven't fully understood what types of non-DID document resources a DID URL can potentially dereference to, if any at all. The PR currently says this:

_"This content MAY be a DID document, a portion of a, a service endpoint, or some other data. The details of the implementation of this function are outside the scope of this specification"_

Is there perhaps some way to rephrase this part to address your concerns? I do think we need to define dereferencing for the reasons stated above, but perhaps we can do that in a way that avoids the "other data" language?

**Comment by OR13** [comment]
I'd rather not be tackling dereferencing and resolution at the same time, I feel like we should focus on PRs not marked as Draft that address DID Resolution first.

**Comment by peacekeeper** [comment]
@jandrieu 

> Second, because dereferencing is still an unsolved puzzle.

I would say it's a _partially_ unsolved puzzle. We see a lot of DID URLs such as `did:ex:123#keys-1` out there. Resolving the DID, and looking up the public key by fragment, that is dereferencing. So I don't think the simple fact that there are DID URLs that can be dereferenced, or the contract for doing that, should be problematic. But I do agree that certain parts of dereferencing, e.g. when it involves paths or query strings, _could_ be problematic.

Just wanted to make this distinction.

**Comment by msporny** [comment]
Closing since the group is now converging on attempting to get PRs #295, #296, #297, #298, #299, and #300 into the spec.

**Comment by jricher** [comment]
PRs #295, #296, #297, #298, #300, and #299 do not cover or include this material.

**Comment by msporny** [comment]
> PRs #295, #296, #297, #298, #300, and #299 do not cover or include this material.

Hmm, right... overzealous closing of PRs in an attempt to streamline to yours.

I'm going to leave this closed as your other PR  that you have requested be left open has Dereferencing in it, and we don't need two separate PRs (at this point) doing so.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-03-11-did-topic#section3) on 2021-03-11

List of resolutions:

- ***[Resolution No. 2](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-03-11-did-topic#resolution2):  Remove the requirement in DID Spec Registries process to register any schema language (e.g., JSON Schema, SHACL, SHeX, CDDL) as a part of property registrations. The process will encourage providing schema languages for appropriate representations.***

<details><summary><i>View the transcript</i></summary>

### 3. Registration process details 

_See github issue [#264](https://github.com/w3c/did-core/issues/264)._




**Manu Sporny:** Orie, what do you feel is the next big set of decisions we need to make? i know there is one around the process. there is the new CDDL requirement.  

> *Manu Sporny:* The Registration Process currently states: "Properties defined in the DID Core Registries MUST include a Concise Data Definition Language (CDDL) [RFC8610] of the Property and its Abstract Data Model structure."

**Manu Sporny:** ^ that got in a while ago. now to register things, we require you to do a JSON-LD context and CDDL. we don't say anything about JSON Schema. My concern is that there are many more people who know JSON Schema than CDDL.  
… do we even want to require CDDL. anything higher priority?  

**Orie Steele:** i don't think so; that is the biggest one  
… This issue comes up as we think about why do we register things. we register things because we want interoperability, to see it is registered, so that people don't invent new ways of doing the same thing. we want people to register stuff so they don't do slight variations of the same functionality. massively redundant. if difficult, they won't do it.  
… The purpose of registration is to help with the representations. because we have multiple representations. we needed a way to register things so we could see how your new registered feature works with those representations  
… JSON-LD context is not helpful for representations other than JSON-LD  
… problem where we can make the bar increase for each representation. increase the number of examples required  
… would block people from registering anything  
… if don't want people to register stuff, asking them to support all representations in their registrations is a good way. problem since beginning of WG, need to address it.  
… Proposal: for every registered representation that is defined in the core representation, you must provide some evidence to show that new properties work. only need to require examples. structural. non-core representations not required to support.  
… willing to hear other proposals  

**Brent Zundel:** can you give us the headline of the topic and we will come back to it  

**Ivan Herman:** want to separate two things. i think i understand and agree with what Orie said. for rep not defined in core spec, we sort of put them aside in terms of the reg process. question: how high is the bar today, and will it become a problem that any new properties that i introduce requires me to have JSON-LD and CDDL (and JSON Schema?) to become a registration? see reason, to give interoperability, but don't know if this will become an obstacle for people to register properties.  

> *Manu Sporny:* [https://w3c.github.io/did-spec-registries/#controller](https://w3c.github.io/did-spec-registries/#controller)

**Manu Sporny:** looking at the registries right now. for every property, we have three properties. the normative definition. strongly defensible. where do you define it. makes sense.  
… JSON-LD context: kind of makes sense. nice to have a context. but JSON folks might not really care. an argument for that.  
… where the WG got to is that they are so class, might as well have it  
… when it comes to schema languages: they are largely representation-specific. if have a json schema, say DIDs must be strings. if you try to apply against a CBOR-LD encoded DID document, it would fail. because CBOR-LD does semantic compression. would not pass CDDL for pure CBOR.  
… now into a position where we are saying that people have to provide schema checks against potentially every representation - at least the ones they care about. that is a very high bar  
… number of people who can right JSON-LD: less than JSON. CDDL: even less  
… My argument to not include any schema checking in the registries. that's not their purpose. in the registration, could have a see also section where people can list the schemas. or have it associated with the normative definition. purely optional, not required for registration  
… because we are trying to get people to register stuff. low bar.  

> *Orie Steele:* +1 manu

**Drummond Reed:** i want to second the idea of both orie and manu that a registration should supply the appropriate schema  
… to support representations should supply the necessary detail in the associatd spec. must have spec and it must meet certain requirements. but agree it shouldn't be a requirement to produce a mapping that you have no interest in. should be encouraged.  

> *Manu Sporny:* +1 to what Drummond just said.

> *Orie Steele:* +1

**Jonathan Holt:** to tote benefits of CDDL. too much credit that i learned it overnight, it's not that hard. it is a concise data definition language. json schema version 7 still draft. not normative yet. CDDL is a normatively described extension of ABNF rules. can right ABNF in CDDL. phenomenally expressive, and normative.  
… I understand the bar. having written JSON Schema for the VC spec, chasing my tail, it was moving so quickly, to figure out what was in or not in. JSON Schema was not the right approach there.  
… CDDL a superset of JSON  
… both JSON-LD as well as CBOR-LD are an expansion algorithm. taking the external dictionary of CBOR-LD, expanding it as you would in JSON-LD  
… get an RDF expansion. that is what needs to be modelled  

> *Drummond Reed:* Just to be clear, I am a big CDDL fan. I just don't think we need to *require* all new registry entries to provide mappings to all representations. We need to *encourage* that.

**Jonathan Holt:** substitute the external dictionary, to get the same #1 means controller in the registry  
… how to get the expressivity of the registry  

> *Manu Sporny:* Agree, CDDL is great -- if you have a community that knows how to use it and a representation that uses it -- we're not in that position.

> *Markus Sabadello:* +1 to ivan

**Ivan Herman:** first of all, to react directly to what jonathan said. i don't think the issue here is whether CDDL is a good thing or bad thing. i think it is a good thing. but that is a different point. in the registry, the big difference between the context and JSON-LD is that for JSON-LD to work with a given property, you _must_ have a way to say that this property maps onto that URL, because this is what JSON-LD context is all about. essentially the only thing it does is tell you what thing maps to what URL.  

> *Amy Guy:* +1 to ivan.. context is for semantic interop, not schema validation

> *Orie Steele:* +1 ivan

**Ivan Herman:** all the other things are good things to have, help implementors. good citizen may want to do that as part of specification of property. but there have to draw line at what we require, and what recommend for good citizens to do  

> *Manu Sporny:* +1 to what Ivan just said.

> *Drummond Reed:* It is true that we agreed in Amsterdam that a JSON-LD context must be provided.

**Markus Sabadello:** +1. also what we said when decided to introduce the ADM and different representations, i thought that was the reason we had the resolution to introduce those things at the same time, and any newly registered property must provide a JSON-LD context, for the reasons ivan mentioned.  

**Drummond Reed:** i do think ivan has a good point. it also feels to me like the lightest of all the requirements, to just make sure that there is a mapping to unambiguous URLs for anything being proposed.  
… want to ask, is that a proposal we are making here?  
… asking for clarification. thought ivan was saying should still have JSON-LD requirement  

**Ivan Herman:** JSON-LD is not a schema language  

**Drummond Reed:** ok  

**Ivan Herman:** that is the mistake that was done in earlier discussion. schema language is shacl or shex for rdf, _not_ the context file  

**Drummond Reed:** want to make modification that we encourage authors to do that  

> **Proposed resolution: Remove the requirement in DID Spec Registries process to register any schema language (e.g., JSON Schema, SHACL, SHeX, CDDL) as a part of property registrations. The process will encourage providing schema languages for appropriate representations.** *(Manu Sporny)*

> *Orie Steele:* +1

> *Ivan Herman:* +1

> *Drummond Reed:* +1

> *Ted Thibodeau Jr.:* +1

> *Amy Guy:* +1

> *Charles Lehner:* +1

> *Brent Zundel:* +1

> *Manu Sporny:* +1

> *Markus Sabadello:* +1

> *Amy Guy:* "relevant" representations

**Manu Sporny:** appropriate representations: for the languages for which they are appropriate  

> ***Resolution #2: Remove the requirement in DID Spec Registries process to register any schema language (e.g., JSON Schema, SHACL, SHeX, CDDL) as a part of property registrations. The process will encourage providing schema languages for appropriate representations.***

**Brent Zundel:** always good to make our editors happy  

**Drummond Reed:** reading through it... what is the rule around JSON-LD contexts, or do we need one?  

> *Manu Sporny:* [https://w3c.github.io/did-spec-registries/#the-registration-process](https://w3c.github.io/did-spec-registries/#the-registration-process)

**Ivan Herman:** that stays, is not changed  

> *Orie Steele:* See: [https://github.com/w3c/did-spec-registries/issues/264](https://github.com/w3c/did-spec-registries/issues/264)

**Drummond Reed:** editorial change: we should always number all the way down the list  

**Brent Zundel:** 12 minutes left on call  

**Drummond Reed:** proposal Orie editor-in-chief for life  

**Brent Zundel:** topics to bring up?  

</details>

File: index.html [hunk 1]
```diff
@@ -2706,13 +2706,105 @@ <h2>
...
+                            conform to valid syntax. (See <a href="#did-url-syntax"></a>.)</dd>
+                    </dl>
+
+                </dd>
+
+            </dl>
+        </p>
+
+        <p>
+The <code>content</code> of this function's output is a byte stream of data. This byte stream MAY
```
[the remark this answers is not in the document]
Comment (TallTed) [reply index.html:2864]: If this gets revived somewhere, `The <code>resource</code> of this function's output ...` would read better as `The <code>resource</code> output by this function ...`.

**Pull request closed without merging**